File: server/services/drawdownCalculator.ts
/**
 * drawdownCalculator.ts
 * ─────────────────────────────────────────────────────────────────────────────
 * Thin Node.js bridge between the Express route layer and the Python drawdown
 * analysis engine (server/python/drawdown_calculator.py).
 *
 * RESPONSIBILITIES (and ONLY these):
 *   1. Accept raw trade rows + optional startingBalance from the route handler
 *   2. Serialise the payload to JSON and pipe it to the Python process via stdin
 *   3. Collect stdout, parse the JSON result, and return it to the caller
 *   4. Handle every failure mode (spawn error, timeout, bad JSON, non-zero exit)
 *      and resolve — never reject — with a typed DrawdownResult
 *
 * All drawdown computation logic lives exclusively in:
 *   server/python/drawdown_calculator.py       (orchestrator)
 *   server/python/drawdown_core.py             (peak/trough/recovery)
 *   server/python/drawdown_streaks.py          (consecutive loss streaks)
 *   server/python/drawdown_heatmap.py          (pair×strategy heatmap)
 *   server/python/drawdown_distribution.py     (size distribution histogram)
 *   server/python/drawdown_frequency.py        (frequency by attribute)
 *   server/python/drawdown_metrics.py          (KPI summary)
 *   server/python/drawdown_structural.py       (structural failure analysis)
 *   server/python/drawdown_sessions.py         (per-session comparison)
 * Do NOT add calculation logic here.
 *
 * USAGE:
 *   import { computeDrawdown } from './services/drawdownCalculator';
 *   const result = await computeDrawdown(trades, session.startingBalance);
 *
 * ROUTE THAT CALLS THIS:
 *   GET /api/drawdown/compute?sessionId=X
 *   Defined in server/routes.ts — fetches session trades from storage, then
 *   calls computeDrawdown() and returns the result as JSON.
 * ─────────────────────────────────────────────────────────────────────────────
 */

import { spawn, ChildProcess } from "child_process";
import * as path from "path";

// ── Constants ─────────────────────────────────────────────────────────────────

const PYTHON_SCRIPT_PATH = path.join(
  process.cwd(),
  "server",
  "python",
  "drawdown_calculator.py"
);

const TIMEOUT_MS = 30_000;
const PYTHON_BIN = process.env.PYTHON_BIN ?? "python3";

// ── Types ─────────────────────────────────────────────────────────────────────

export interface DrawdownResult {
  success: boolean;
  topStats?: {
    maxDrawdown: number;
    avgDrawdown: number;
    recoveryFactor: number;
    trendAlignment: number;
  };
  streaks?: Record<string, any>;
  heatmap?: any[];
  distribution?: Record<string, any>;
  frequency?: Record<string, any>;
  structural?: Record<string, any>;
  sessionComparison?: any[];
  error?: string;
}

// ── Public API ────────────────────────────────────────────────────────────────

/**
 * computeDrawdown
 * ───────────────
 * Spawns the Python drawdown engine, feeds it the trade array via stdin, and
 * returns the parsed analysis result. Always resolves — never rejects.
 *
 * @param trades           Raw journal entry rows from the database
 * @param startingBalance  Optional account starting balance for recovery calculations
 */
export async function computeDrawdown(
  trades: any[],
  startingBalance?: number
): Promise<DrawdownResult> {
  return new Promise((resolve) => {
    if (!Array.isArray(trades)) {
      resolve({ success: false, error: "trades must be an array" });
      return;
    }

    const payload = { trades, ...(startingBalance !== undefined && { startingBalance }) };

    const child: ChildProcess = spawn(PYTHON_BIN, [PYTHON_SCRIPT_PATH], {
      env: { ...process.env },
    });

    let stdout = "";
    let stderr = "";

    const timer = setTimeout(() => {
      try { child.kill("SIGTERM"); } catch {}
      resolve({ success: false, error: "Drawdown computation timed out (30s)" });
    }, TIMEOUT_MS);

    child.stdin?.write(JSON.stringify(payload));
    child.stdin?.end();

    child.stdout?.on("data", (chunk) => { stdout += chunk.toString(); });
    child.stderr?.on("data", (chunk) => { stderr += chunk.toString(); });

    child.on("close", (code) => {
      clearTimeout(timer);
      if (stderr.trim()) {
        console.error("[DrawdownCalculator] Python stderr:", stderr.slice(0, 500));
      }
      if (stdout.trim()) {
        try {
          resolve(JSON.parse(stdout.trim()) as DrawdownResult);
          return;
        } catch {
          console.error("[DrawdownCalculator] JSON parse error:", stdout.slice(0, 300));
        }
      }
      resolve({
        success: false,
        error: code !== 0 ? `Python exited with code ${code}: ${stderr}` : "No output from drawdown engine",
      });
    });

    child.on("error", (err) => {
      clearTimeout(timer);
      console.error("[DrawdownCalculator] Spawn error:", err);
      resolve({ success: false, error: err.message });
    });
  });
}

File: server/python/drawdown_calculator.py
"""
drawdown_calculator.py
──────────────────────────────────────────────────────────────────────────────
Main entry point for the drawdown analysis engine. This script is spawned by
the Node.js bridge (server/services/drawdownCalculator.ts) as a child process.

HOW IT WORKS:
  - Receives a JSON payload via stdin: { "trades": [...], "startingBalance": float }
  - Orchestrates all individual drawdown sub-modules:
      drawdown_core.py         → peak/trough/recovery periods, max drawdown %
      drawdown_streaks.py      → consecutive loss streaks, streak depth analysis
      drawdown_heatmap.py      → pair×strategy heatmap matrix of loss severity
      drawdown_distribution.py → histogram distribution of drawdown sizes
      drawdown_frequency.py    → frequency breakdown by attribute (session, strategy, psychology)
      drawdown_metrics.py      → summary KPIs: avg drawdown, recovery factor, trend alignment
      drawdown_structural.py   → structural failure analysis (context, session, risk)
      drawdown_sessions.py     → per-session drawdown comparison
  - Aggregates results into a single JSON object and writes it to stdout
  - Writes any errors to stderr and exits with code 1

OUTPUT SHAPE (written to stdout):
{
  "success": true,
  "topStats": {
    "maxDrawdown": float,          // e.g. -8.42 (percentage)
    "avgDrawdown": float,          // average drawdown % across all drawdown periods
    "recoveryFactor": float,       // net profit / max drawdown
    "trendAlignment": float        // % of trades taken in trend direction
  },
  "streaks": {
    "longestLossStreak": int,
    "worstStreakDepth": float,     // cumulative % lost in worst streak
    "avgRecoveryTrades": float,    // avg trades needed to recover after streak
    "streakHistory": [             // list of all losing streaks found
      { "start": int, "end": int, "length": int, "depth": float }
    ]
  },
  "heatmap": [                     // pair × strategy loss severity matrix
    {
      "pair": str,
      "cells": [
        { "strategy": str, "val": float, "trades": int, "losses": int }
      ]
    }
  ],
  "distribution": {                // histogram of drawdown sizes
    "buckets": [str],              // e.g. ["0-1%", "1-2%", ...]
    "counts": [int]
  },
  "frequency": {                   // loss frequency by attribute category
    "attr": [
      { "cat": str, "name": str, "total": int, "losses": int }
    ],
    "instr": [...],
    "session": [...]
  },
  "structural": {                  // structural failure context breakdown
    "context": [...],
    "session": [...],
    "risk": [...]
  },
  "sessionComparison": [           // per-session drawdown comparison
    { "sessionId": int, "sessionName": str, "maxDrawdown": float, "recoveryFactor": float }
  ],
  "error": null
}

NOTES:
  - All percentage values are expressed as negative floats (e.g. -8.42 means -8.42%)
  - Trade rows use the same camelCase schema as journal_entries table in shared/schema.ts
  - If fewer than 2 trades exist, return success=true with zeroed/empty stats
  - Never print anything other than the final JSON to stdout (use sys.stderr for logs)
"""

import sys
import json

def main():
    raw = sys.stdin.read()
    try:
        payload = json.loads(raw)
    except json.JSONDecodeError as e:
        print(json.dumps({"success": False, "error": f"Invalid JSON input: {e}"}))
        sys.exit(1)

    trades = payload.get("trades", [])
    starting_balance = payload.get("startingBalance")

    # TODO: import and call each sub-module, then aggregate results:
    # from drawdown_core import compute_core_drawdown
    # from drawdown_streaks import compute_streaks
    # from drawdown_heatmap import compute_heatmap
    # from drawdown_distribution import compute_distribution
    # from drawdown_frequency import compute_frequency
    # from drawdown_metrics import compute_summary_metrics
    # from drawdown_structural import compute_structural
    # from drawdown_sessions import compute_session_comparison

    result = {
        "success": True,
        "topStats": {},
        "streaks": {},
        "heatmap": [],
        "distribution": {},
        "frequency": {},
        "structural": {},
        "sessionComparison": [],
        "error": None
    }

    print(json.dumps(result))

if __name__ == "__main__":
    main()
